src/Sylius/Behat/Context/Domain/PromotionContext.php

```diff
@@ -0,0 +1,111 @@
+use Sylius\Component\Core\Test\Services\SharedStorageInterface;
+use Sylius\Component\Promotion\Model\CouponInterface;
+use Sylius\Component\Promotion\Repository\PromotionRepositoryInterface;
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+
+/**
+ * @author Jan Góralski <jan.goralski@lakion.com>
+ */
+final class PromotionContext implements Context
+{
+    /**
+     * @var SharedStorageInterface
+     */
+    private $sharedStorage;
+
+    /**
+     * @var PromotionRepositoryInterface
+     */
+    private $promotionRepository;
+
+    /**
+     * @var RepositoryInterface
+     */
+    private $couponRepository;
+
+    /**
+     * @param SharedStorageInterface $sharedStorage
+     * @param PromotionRepositoryInterface $promotionRepository
+     * @param RepositoryInterface $couponRepository
+     */
+    public function __construct(
+        SharedStorageInterface $sharedStorage,
+        PromotionRepositoryInterface $promotionRepository,
+        RepositoryInterface $couponRepository
+    ) {
+        $this->sharedStorage = $sharedStorage;
+        $this->promotionRepository = $promotionRepository;
+        $this->couponRepository = $couponRepository;
+    }
+
+    /**
+     * @When /^I try to delete ("([^"]+)" coupon)$/
+     */
+    public function iTryToDeleteCoupon(CouponInterface $coupon)
+    {
+        try {
+            $this->couponRepository->remove($coupon);
+
+            throw new \Exception(sprintf('Coupon "%s" has been removed, but it should not.', $coupon->getCode()));
+        } catch(ForeignKeyConstraintViolationException $exception) {
+            $this->sharedStorage->set('last_exception', $exception);
+        }
+    }
+
+    /**
+     * @When /^I delete ("([^"]+)" coupon)$/
+     */
+    public function iDeleteCoupon(CouponInterface $coupon)
+    {
+        $this->sharedStorage->set('coupon_id', $coupon->getId());
+        $this->couponRepository->remove($coupon);
+    }
+
+    /**
+     * @Then /^([^"]+ should no longer) exist in the registry$/
+     */
+    public function couponShouldNotExistInTheRegistry($couponId)
+    {
+        expect($this->couponRepository->find($couponId))->toBe(null);
+    }
+
+    /**
+     * @Then I should be notified that it is in use and cannot be deleted
+     */
+    public function iShouldBeNotifiedOfFailure()
+    {
+        expect($this->sharedStorage->get('last_exception'))
+            ->toThrow(ForeignKeyConstraintViolationException::class)
+        ;
+    }
+
+    /**
+     * @Then /^([^"]+) should still exist in the registry$/
+     */
+    public function couponShouldStillExistInTheRegistry(CouponInterface $coupon)
+    {
+        expect($this->couponRepository->find($coupon->getId()))->toNotBe(null);
+    }
+
+    /**
+     * Not applicable in the domain scope
+     *
+     * @Then I should be notified that it has been successfully deleted
+     */
+    public function iShouldBeNotifiedOfSuccess(){}
```

Comment:
IIRC we write the empty functions like:
```php
/**
 * Not applicable in the domain scope
 * 
 * @Then I should be notified that it has been successfully deleted
 */
public function iShouldBeNotifiedOfSuccess()
{
}
```

IMHO it would be even more readable if we put the comment inside the method like:
```php
/**
 * @Then I should be notified that it has been successfully deleted
 */
public function iShouldBeNotifiedOfSuccess()
{
    // Not applicable in the domain scope
}
```